### DISCUSSION TITLE
Detect line endings in `next-env.d.ts`

### DISCUSSION THREAD

**Discussion by whitneyit** [vercel/next.js#27039] [Ideas]
### Describe the feature you'd like to request

The [`writeAppTypeDeclarations()`](https://github.com/vercel/next.js/blob/d60c95cb57497396d534eb8dd6fce9a192fc22fe/packages/next/lib/typescript/writeAppTypeDeclarations.ts) function uses `os.EOL` to write new lines to the file.

If a developer on a Windows machine uses a file committed by a developer on a Mac for instance, this causes an unnecessary change to the file.

### Describe the solution you'd like

Perhaps the [`detect-newline`](https://www.npmjs.com/package/detect-newline) package could be used to ensure that if the file exists, the appropriate newline is being used.

Considering that there are only a handful of [references to `os.EOL`](https://github.com/vercel/next.js/search?q=os.eol) in the codebase, perhaps this approach could be applied more generally across the project?

### Describe alternatives you've considered

Aside from running `git checkout next-env.d.ts` there is no real alternative 🤷

**Comment by timneutkens** [comment]
Would #27028 solve your case?

**Reply by whitneyit** [reply]
I don't think so. That seems to only address the trailing EOL.

**Reply by whitneyit** [reply]
I would expect a solution would detect the given newline in the file and use that.

Reason being, the detected EOL may differ from the OS's EOL.

**Reply by awwit** [reply]
In our project, the same problem arose, developers with different operating systems find themselves in a situation where the file is overwritten when starting next (due to different endings of the line).
And the `next-env.d.ts` file always gets in the commit.

**Reply by LuudJanssen** [reply]
I'm also facing this issue. As far as I can see there a couple of approaches to fix this:

1. Don't regenerate the `next-env.d.ts` file if the contents are the same as before. Right now, only the line endings change for me everytime I run `next dev` or `next build`, the rest of the contents stay the same.
2. Check for a `.editorconfig` `.gitattributes` or `.prettierrc` file to check if there is a configuration that dictates the line endings for a project. In most of my projects I have at least one of them set to `LF`.
3. Use a tool like `detect-newline`, which @whitneyit suggested to scan the previous line endings in the `next-env.d.ts` file and use those in the generation of a new `next-env.d.ts` file.
4. Since `next-env.d.ts` is auto-generated, recommend users to add it to their `.gitignore` file.

If you need any additional manpower and an example of a Windows user to test this, I'm available to pick up this issue and provide a PR.

**Comment by ZoranRavic** [comment]
If `next-env.d.ts` is always auto generated, then why isn't it added to `.gitignore` by default?

**Reply by Wattyyy** [reply]
@ZoranRavic 
I have thought the same thing, so I opened the #38541  for adding `next-env.d.ts` to the `.gitignore` template.
